en/02_Developer_Guides/03_Forms/07_Using_GridField_With_Arbitrary_Data.md

```diff
@@ -0,0 +1,437 @@
+];
+// Pass the BasicSearchContext into the GridFieldFilterHeader component
+$searchContext->setFields(FieldList::create($searchFields));
+$gridField->getConfig()->getComponentByType(GridFieldFilterHeader::class)->setSearchContext($searchContext);
+```
+
+### Exporting data {#arraydata-export}
+
+If you want to export or print your data, you don't have to do anything special - just make sure to include the `GridFieldExportButton` and `GridFieldPrintButton` components.
+
+[info]
+These will both use the field list you [passed into `GridFieldDataColumns`](#arraydata-display-as-rows) to know which fields they should use - though you can explicitly call [`GridFieldExportButton::setExportColumns()`](api:SilverStripe\Forms\GridField\GridFieldExportButton::setExportColumns()) and [`GridFieldPrintButton::setPrintColumns()`](api:SilverStripe\Forms\GridField\GridFieldPrintButton::setPrintColumns()) if you want to export/print different columns than those displayed in the grid view.
+[/info]
+
+```php
+use SilverStripe\Forms\GridField\GridFieldExportButton;
+use SilverStripe\Forms\GridField\GridFieldPrintButton;
+$gridField->getConfig()->addComponents([
+    GridFieldExportButton::create('buttons-before-left'),
+    GridFieldPrintButton::create('buttons-before-left'),
+]);
+```
+
+### Viewing data in a form {#arraydata-view}
+
+For data to be viewed in a read-only form, each record in the list must have an `ID` field, and the value of that field must be a positive integer. This is used in the URL for the form. Without it, the gridfield has no way to know which record it should be displaying in the form.
+
+You'll need to add a `GridFieldDetailForm` component to the `GridField` and tell it how to represent your data by passing a [`FieldList`](api:SilverStripe\Forms\FieldList) into [`GridFieldDetailForm::setFields()`](api:SilverStripe\Forms\GridField\GridFieldDetailForm::setFields()).
+
+[hint]
+Because `ArrayData` doesn't implement a `canEdit()` method, the form will be implicitly turned into a read-only form for you. You don't need to worry about passing in read-only form fields.
+[/hint]
+
+```php
+use SilverStripe\Forms\FieldList;
+use SilverStripe\Forms\GridField\GridFieldDetailForm;
+use SilverStripe\Forms\GridField\GridFieldViewButton;
+use SilverStripe\Forms\HiddenField;
+use SilverStripe\Forms\TextField;
+
+$detailForm = GridFieldDetailForm::create();
+$detailForm->setFields(FieldList::create([
+    HiddenField::create('ID'),
+    TextField::create('FeldName', 'View Field Label'),
+]));
+$gridField->getConfig()->addComponents([
+    GridFieldViewButton::create(),
+    $detailForm,
+]);
+```
+
+## Representing data in your own class
+
+As mentioned in the preamble above, the class representing your data must be a subclass of `ViewableData` in order for it to be used in a `GridField`. It's possible that some components will work even if your class isn't a subclass of `ViewableData`, but that use case is not supported and could break at any time.
```

Comment:
As a general principle, you should write each section without assuming that people have read all the previous bit.

I also don't think we should muddy the water by implying that if you choose not to use ViewableData, bits might still be functional.

```suggestion
The class representing your data must be a subclass of `ViewableData` in order for it to be used in a `GridField`.
```

Reply:
I'm not assuming they've read it. That's why I'm explicitly mentioning the part above - that way people will know "Ahh, this has been mentioned before, I should probably go read that part as well as it may have some relevant information that isn't being duplicated here"

I don't want to make this suggested change, because pointing out there's more related information above seems useful to me.

I've removed the bit about non-viewabledata.

Reply:
To the extent that we think previous information is relevant, then we should have a same page link to it then.

Reply:
Added a link